Expose and warn about Node Filters in Scene Tree Dock

Mickeon opened the pull request:
Continuation of #65352. (see also https://github.com/godotengine/godot/pull/65352#pullrequestreview-1100948259)

This PR adds "**Filter by Type**" and "**Filter by Group**"  in the Scene Tree Dock's **MenuButton** (the one on the top right featuring 3 dots), exposing them to everyone that doesn't _religiously_ look at this engine's source code.

| Menu | Tooltip |
| :---: | :---: |
| ![image](https://user-images.githubusercontent.com/66727710/190711678-2fd4f2da-d8d7-4602-bf81-9e2db7b7f03c.png) | ![image](https://user-images.githubusercontent.com/66727710/190712072-a4b6656a-353f-4315-9e11-e2330b09d56e.png)|
| |  Hovering on them displays an useful tooltip. |

| Behavior | Warning |
| :---: | :---: |
| ![Showcase](https://i.gyazo.com/b9ecf539cfa1e88223032ef31ce5da3b.gif)  | ![image](https://user-images.githubusercontent.com/66727710/193866518-d213f97a-5ef7-4bef-8146-a1b79cc65fef.png) |
| When clicking on either of these, the matching parameter is appended to the terms, and the caret is automatically brought to the end. | When typing a filter that cannot be identified, a warning icon is displayed. The reason is explained in the tooltip. Before this, the temporary solution of #65352 is to spam warnings to the console. 

And lastly, the same options are also quickly available by **right-clicking** or **middle-clicking** in the text field.
| Right Click | Middle Click |
| :---: | :---: |
| ![image](https://user-images.githubusercontent.com/66727710/190725688-4a135fd2-115f-4236-a062-fa3f800f8094.png) | ![MiddleClick](https://user-images.githubusercontent.com/66727710/190724570-81f49494-52ec-4c22-afb1-bc6c69e726a9.png) |

Mickeon committed:
Expose and warn about Node Filters in Scene Tree Dock

Adds "Filter by Type" and "Filter by Group" in the Scene Tree Dock's MenuButton.
Hovering on them displays an useful tooltip.
When selecting these items, the matching parameter is appended to the terms, and the caret is automatically brought to the end.

When typing a filter that cannot be identified, a warning icon is displayed. The reason is explained as a tooltip.

The same options are also quickly available by right-clicking or middle-clicking in the text field.

akien-mga commented:
Looks good, but the warning system is weird that it shows a warning icon but the actual warning is shown in the menu. That's not something we do anywhere else in the editor.

The warning should be shown in a tooltip on hover of the warning icon, like for node configuration warnings.

akien-mga commented on editor/scene_tree_dock.cpp:
```diff
@@ -2922,11 +2924,88 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+			p_menu->set_item_text(idx, warning);
+			p_menu->set_item_icon(idx, get_theme_icon(SNAME("Warning"), SNAME("EditorIcons")));
+			p_menu->set_item_indent(idx, -1);
+		};
```
Unnecessary semicolon.

Mickeon commented:
> The warning should be shown in a tooltip on hover of the warning icon, like for node configuration warnings.

Not sure it would be possible to have the tooltip display exactly there, without resulting into something poor and hacky. I want to note that the Warning icon in this PR is actually still the "clear" TextEdit button. Its behavior remains unchanged, which means it would clear the text on click instead of telling the user what's wrong. Not the worst thing in the world, really, just putting it out there.

Mickeon commented:
Going to update this PR to display the warning as a tooltip hovering over the filter TextEdit:
![image](https://user-images.githubusercontent.com/66727710/193866518-d213f97a-5ef7-4bef-8146-a1b79cc65fef.png)

Mickeon commented:
All good now, I hope.

KoBeWi commented:
I have nodes without any groups. With empty `group:` filter, some of them disappear:
![godot windows editor dev x86_64_8npSWHobZk](https://user-images.githubusercontent.com/2223172/194135203-65c036ed-2ed2-44c4-b4e5-0fee8d3a00b6.gif)
Why is that?

Mickeon commented:
Thank you for the testing as always, @KoBeWi. The group thing may not be related to this PR, but it'd still be nice to see if something can be done about it, because I implemented the feature and it'd be nice to address it. It may be related to the way groups are detected. It'd love to take a look at that project you're testing it with to spot the issue.

Meanwhile, I addressed the right-click options not working.

Mickeon commented:
Ok I can address it in another PR. Perhaps within https://github.com/godotengine/godot/pull/65939 to make it quick.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2883,11 +2885,84 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+	}
+
+	if (mb->is_pressed() && mb->get_button_index() == MouseButton::MIDDLE) {
+		filter_quick_menu->clear();
```
From what I see, this menu is always the same, so I think there is no reason to clear it each time.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2883,11 +2885,84 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
 void SceneTreeDock::_update_tree_menu() {
 	PopupMenu *tree_menu = button_tree_menu->get_popup();
-	tree_menu->set_item_checked(tree_menu->get_item_idx_from_text(TTR("Auto Expand to Selected")), EditorSettings::get_singleton()->get("docks/scene_tree/auto_expand_to_selected"));
+	tree_menu->clear();
```
This also doesn't need to be cleared. Clearing is only useful when list of items can change.

Mickeon replied:
Right. This used to change before the warning message was turned into a tooltip. A lot of these initialisations on the bottom page are really messy, but it is what it is.

As I was writing this, however, I realised that by fixing this, icons do not load. Is it a good idea to not clear the filter menus in favour of writing more code around to ensure the menu items are set on NOTIFICATION_READY and the icons are always up to date?

KoBeWi replied:
Eh, if it would require more code then maybe it's ok this way. It affects performance, but not in a meaningful amount.

akien-mga commented:
Thanks!
